cpp/src/join/full_join.cuh

```diff
@@ -0,0 +1,172 @@
+    rmm::device_vector<index_type> invalid_index_map(right_table_row_count, 1);
+    //Functor to check for index validity since left joins can create invalid indices
+    valid_range<size_type> valid(0, right_table_row_count);
+
+    //invalid_index_map[index_ptr[i]] = 0 for i = 0 to right_table_row_count
+    //Thus specifying that those locations are valid
+    thrust::scatter_if(
+        rmm::exec_policy(stream)->on(stream),
+        thrust::make_constant_iterator(0),
+        thrust::make_constant_iterator(0) + right_indices.size(),
+        right_indices.begin<index_type>(),//Index locations
+        right_indices.begin<index_type>(),//Stencil - Check if index location is valid
+        invalid_index_map.begin(),//Output indices
+        valid);//Stencil Predicate
+    size_type begin_counter = static_cast<size_type>(0);
+    size_type end_counter = static_cast<size_type>(right_table_row_count);
+
+    //Create list of indices that have been marked as invalid
+    indices_count = thrust::copy_if(
+        rmm::exec_policy(stream)->on(stream),
+        thrust::make_counting_iterator(begin_counter),
+        thrust::make_counting_iterator(end_counter),
+        invalid_index_map.begin(),
+        unmatched_indices_ptr,
+        thrust::identity<index_type>()) -
+      unmatched_indices_ptr;
+  }
+  return std::make_unique<column>(
+      data_type(cudf::experimental::type_to_id<index_type>()),
+      indices_count,
+      std::move(unmatched_indices));
+}
+
+
+/* --------------------------------------------------------------------------*/
+/**
+* @Synopsis  Creates a table containing the complement of left join indices.
+* This table has two columns. The first one is filled with JoinNoneValue(-1)
+* and the second one contains values from 0 to right_table_row_count - 1
+* excluding those found in the right_indices column.
+* 
+* @Param right_indices Column of indices
+* @Param left_table_row_count Number of rows of left table
+* @Param right_table_row_count Number of rows of right table
+* @param mr Optional, the memory resource that will be used for allocating
+* the device memory for the new column
+* @param stream Optional, stream on which all memory allocations and copies
+* will be performed
+*
+* @Returns  Table containing the left join indices complement
+*/
+/* ----------------------------------------------------------------------------*/
+template <typename index_type>
+std::unique_ptr<experimental::table>
+get_left_join_indices_complement(
+    column_view const& right_indices,
+    size_type left_table_row_count,
+    size_type right_table_row_count,
+    rmm::mr::device_memory_resource* mr,
+    cudaStream_t stream) {
+
+  //Get array of indices that do not appear in r_index_ptr
+  auto right_indices_complement = create_missing_indices<index_type>(
+      right_indices, left_table_row_count, right_table_row_count, mr, stream);
+
+  auto left_invalid_indices = experimental::allocate_like(right_indices_complement->view());
```

Comment:
Can `allocate_like` take `mr` and `stream` as parameters?

Reply:
`experimental::detail::allocate_like` does.

Reply:
allocate_like uses mr. We do not want to pass the user define mr for this call since left_invalid_indices and all the other indices used for gather are not returned to the user.